File: cocos/core/renderer/core/texture-buffer-pool.ts
/*
 Copyright (c) 2020 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

/**
 * @packageDocumentation
 * @hidden
 */

import { getTypedArrayConstructor, Format, FormatInfos, TextureType, TextureUsageBit,
    TextureFlagBit, Texture, TextureInfo, Device, BufferTextureCopy } from '../../gfx';
import { debug } from '../../platform/debug';

export function nearestPOT (num: number): number {
    --num;
    num |= num >> 16;
    num |= num >> 8;
    num |= num >> 4;
    num |= num >> 2;
    num |= num >> 1;
    ++num;
    return num;
}

export interface ITextureBuffer {
    texture: Texture;
    size: number;
    start: number;
    end: number;
}

export interface ITextureBufferHandle {
    chunkIdx: number;
    start: number;
    end: number;
    texture: Texture;
}

export interface ITextureBufferPoolInfo {
    format: Format; // target texture format
    inOrderFree?: boolean; // will the handles be freed exactly in the order of their allocation?
    alignment?: number; // the data alignment for each handle allocated, in bytes
    roundUpFn?: (size: number, formatSize: number) => number; // given a target size, how will the actual texture size round up?
}

function roundUp (n: number, alignment: number) {
    return Math.ceil(n / alignment) * alignment;
}

export class TextureBufferPool {
    private _device: Device;
    private _format = Format.UNKNOWN;
    private _formatSize = 0;
    private _chunks: ITextureBuffer[] = [];
    private _chunkCount = 0;
    private _handles: ITextureBufferHandle[] = [];
    private _region0 = new BufferTextureCopy();
    private _region1 = new BufferTextureCopy();
    private _region2 = new BufferTextureCopy();
    private _roundUpFn: ((targetSize: number, formatSize: number) => number) | null = null;
    private _bufferViewCtor: TypedArrayConstructor = Uint8Array;
    private _channels = 4;
    private _alignment = 1;

    public constructor (device: Device) {
        this._device = device;
    }

    public initialize (info: ITextureBufferPoolInfo) {
        const formatInfo = FormatInfos[info.format];
        this._format = info.format;
        this._formatSize = formatInfo.size;
        this._channels = formatInfo.count;
        this._bufferViewCtor = getTypedArrayConstructor(formatInfo);
        this._roundUpFn = info.roundUpFn || null;
        this._alignment = info.alignment || 1;
        if (info.inOrderFree) { this.alloc = this._McDonaldAlloc; }
    }

    public destroy () {
        for (let i = 0; i < this._chunkCount; ++i) {
            const chunk = this._chunks[i];
            chunk.texture.destroy();
        }
        this._chunks.length = 0;
        this._handles.length = 0;
    }

    public alloc (size: number, chunkIdx?: number) {
        size = roundUp(size, this._alignment);

        let index = -1;
        let start = -1;
        if (chunkIdx !== undefined) {
            index = chunkIdx;
            start = this._findAvailableSpace(size, index);
        }

        if (start < 0) {
            for (let i = 0; i < this._chunkCount; ++i) {
                index = i;
                start = this._findAvailableSpace(size, index);
                if (start >= 0) { break; }
            }
        }

        if (start >= 0) {
            const chunk = this._chunks[index];
            chunk.start += size;
            const handle: ITextureBufferHandle = {
                chunkIdx: index,
                start,
                end: start + size,
                texture: chunk.texture,
            };
            this._handles.push(handle);
            return handle;
        }

        // create a new one
        const targetSize = Math.sqrt(size / this._formatSize);
        const texLength = this._roundUpFn && this._roundUpFn(targetSize, this._formatSize) || Math.max(1024, nearestPOT(targetSize));
        const newChunk = this._chunks[this.createChunk(texLength)];

        newChunk.start += size;
        const texHandle: ITextureBufferHandle = {
            chunkIdx: this._chunkCount - 1,
            start: 0,
            end: size,
            texture: newChunk.texture,
        };
        this._handles.push(texHandle);
        return texHandle;
    }

    public free (handle: ITextureBufferHandle) {
        for (let i = 0; i < this._handles.length; ++i) {
            if (this._handles[i] === handle) {
                this._chunks[handle.chunkIdx].end = handle.end;
                this._handles.splice(i, 1);
                return;
            }
        }
    }

    public createChunk (length: number) {
        const texSize = length * length * this._formatSize;

        debug(`TextureBufferPool: Allocate chunk ${this._chunkCount}, size: ${texSize}, format: ${this._format}`);

        const texture: Texture = this._device.createTexture(new TextureInfo(
            TextureType.TEX2D,
            TextureUsageBit.SAMPLED | TextureUsageBit.TRANSFER_DST,
            this._format,
            length,
            length,
        ));

        const chunk: ITextureBuffer = {
            texture,
            size: texSize,
            start: 0,
            end: texSize,
        };
        this._chunks[this._chunkCount] = chunk;
        return this._chunkCount++;
    }

    public update (handle: ITextureBufferHandle, buffer: ArrayBuffer) {
        const buffers: ArrayBufferView[] = [];
        const regions: BufferTextureCopy[] = [];
        const start = handle.start / this._formatSize;

        let remainSize = buffer.byteLength / this._formatSize;
        let offsetX = start % handle.texture.width;
        let offsetY = Math.floor(start / handle.texture.width);
        let copySize = Math.min(handle.texture.width - offsetX, remainSize);
        let begin = 0;

        if (offsetX > 0) {
            this._region0.texOffset.x = offsetX;
            this._region0.texOffset.y = offsetY;
            this._region0.texExtent.width = copySize;
            this._region0.texExtent.height = 1;

            buffers.push(new this._bufferViewCtor(buffer, begin * this._formatSize, copySize * this._channels));
            regions.push(this._region0);

            offsetX = 0;
            offsetY += 1;
            remainSize -= copySize;
            begin += copySize;
        }

        if (remainSize > 0) {
            this._region1.texOffset.x = offsetX;
            this._region1.texOffset.y = offsetY;

            if (remainSize > handle.texture.width) {
                this._region1.texExtent.width = handle.texture.width;
                this._region1.texExtent.height = Math.floor(remainSize / handle.texture.width);
                copySize = this._region1.texExtent.width * this._region1.texExtent.height;
            } else {
                copySize = remainSize;
                this._region1.texExtent.width = copySize;
                this._region1.texExtent.height = 1;
            }

            buffers.push(new this._bufferViewCtor(buffer, begin * this._formatSize, copySize * this._channels));
            regions.push(this._region1);

            offsetX = 0;
            offsetY += this._region1.texExtent.height;
            remainSize -= copySize;
            begin += copySize;
        }

        if (remainSize > 0) {
            this._region2.texOffset.x = offsetX;
            this._region2.texOffset.y = offsetY;
            this._region2.texExtent.width = remainSize;
            this._region2.texExtent.height = 1;

            buffers.push(new this._bufferViewCtor(buffer, begin * this._formatSize, remainSize * this._channels));
            regions.push(this._region2);
        }

        this._device.copyBuffersToTexture(buffers, handle.texture, regions);
    }

    private _findAvailableSpace (size: number, chunkIdx: number) {
        const chunk = this._chunks[chunkIdx];
        let isFound = false;
        let start = chunk.start;
        if ((start + size) <= chunk.size) {
            isFound = true;
        } else {
            start = 0; // try to find from head again
            const handles = this._handles.filter((h) => h.chunkIdx === chunkIdx).sort((a, b) => a.start - b.start);
            for (let i = 0; i < handles.length; i++) {
                const handle = handles[i];
                if ((start + size) <= handle.start) {
                    isFound = true;
                    break;
                }
                start = handle.end;
            }
            if (!isFound && (start + size) <= chunk.size) {
                isFound = true;
            }
        }
        return isFound ? start : -1;
    }

    // [McDonald 12] Efficient Buffer Management
    private _McDonaldAlloc (size: number) {
        size = roundUp(size, this._alignment);

        for (let i = 0; i < this._chunkCount; ++i) {
            const chunk = this._chunks[i];
            let isFound = false;
            let start = chunk.start;
            if ((start + size) <= chunk.end) {
                isFound = true;
            } else if (start > chunk.end) {
                if ((start + size) <= chunk.size) {
                    isFound = true;
                } else if (size <= chunk.end) {
                    // Try to find from head again.
                    chunk.start = start = 0;
                    isFound = true;
                }
            } else if (start === chunk.end) {
                chunk.start = start = 0;
                chunk.end = chunk.size;
                if (size <= chunk.end) {
                    isFound = true;
                }
            }
            if (isFound) {
                chunk.start += size;
                const handle: ITextureBufferHandle = {
                    chunkIdx: i,
                    start,
                    end: start + size,
                    texture: chunk.texture,
                };
                this._handles.push(handle);
                return handle;
            }
        }

        // create a new one
        const targetSize = Math.sqrt(size / this._formatSize);
        const texLength = this._roundUpFn && this._roundUpFn(targetSize, this._formatSize) || Math.max(1024, nearestPOT(targetSize));
        const newChunk = this._chunks[this.createChunk(texLength)];

        newChunk.start += size;
        const texHandle: ITextureBufferHandle = {
            chunkIdx: this._chunkCount,
            start: 0,
            end: size,
            texture: newChunk.texture,
        };
        this._handles.push(texHandle);
        return texHandle;
    }
}

File: native/cocos/renderer/core/TextureBufferPool.cpp
/****************************************************************************
 Copyright (c) 2021 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "renderer/core/TextureBufferPool.h"
#include <cmath>
#include "core/ArrayBuffer.h"
#include "core/TypedArray.h"

namespace {

uint32_t roundUp(uint32_t n, uint32_t alignment) {
    return static_cast<uint32_t>(std::ceil(n / alignment)) * alignment;
}
} // namespace

namespace cc {

TextureBufferPool::TextureBufferPool(gfx::Device *device) {
    _device = device;
}

void TextureBufferPool::initialize(const ITextureBufferPoolInfo &info) {
    const auto &formatInfo = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(info.format)];
    _format                = info.format;
    _formatSize            = formatInfo.size;
    _channels              = formatInfo.count;
    _roundUpFn             = info.roundUpFn.has_value() ? info.roundUpFn.value() : nullptr;
    _alignment             = info.alignment.has_value() ? info.alignment.value() : 1;
    _useMcDonaldAlloc      = info.alignment.has_value() && info.alignment.value();
}

void TextureBufferPool::destroy() {
    for (auto &chunk : _chunks) {
        CC_SAFE_DESTROY_AND_DELETE(chunk.texture);
    }
    _chunks.clear();
    _handles.clear();
}

ITextureBufferHandle TextureBufferPool::alloc(uint32_t size) {
    if (_useMcDonaldAlloc) {
        return mcDonaldAlloc(size);
    }

    size          = roundUp(size, _alignment);
    index_t index = CC_INVALID_INDEX;
    index_t start = CC_INVALID_INDEX;
    if (start < 0) {
        for (index_t i = 0; i < _chunkCount; ++i) {
            index = i;
            start = findAvailableSpace(size, index);
            if (start >= 0) break;
        }
    }

    if (start >= 0) {
        auto chunk = _chunks[index];
        chunk.start += static_cast<index_t>(size);
        ITextureBufferHandle handle;
        handle.chunkIdx = index;
        handle.start    = start;
        handle.end      = static_cast<index_t>(start + size);
        handle.texture  = chunk.texture;

        _handles.emplace_back(handle);
        return handle;
    }
    // create a new one
    auto     targetSize = static_cast<int32_t>(std::sqrt(size / _formatSize));
    uint32_t texLength  = _roundUpFn ? _roundUpFn(targetSize, _formatSize) : std::max(1024, static_cast<int>(utils::nextPOT(targetSize)));
    auto     newChunk   = _chunks[createChunk(texLength)];

    newChunk.start += static_cast<index_t>(size);
    ITextureBufferHandle texHandle;
    texHandle.chunkIdx = static_cast<index_t>(_chunkCount - 1);
    texHandle.start    = 0;
    texHandle.end      = static_cast<index_t>(size);
    texHandle.texture  = newChunk.texture;
    _handles.emplace_back(texHandle);
    return texHandle;
}

ITextureBufferHandle TextureBufferPool::alloc(uint32_t size, index_t chunkIdx) {
    size          = roundUp(size, _alignment);
    index_t index = chunkIdx;
    index_t start = findAvailableSpace(size, index);

    if (start < 0) {
        for (index_t i = 0; i < _chunkCount; ++i) {
            index = i;
            start = findAvailableSpace(size, index);
            if (start >= 0) break;
        }
    }

    if (start >= 0) {
        auto chunk = _chunks[index];
        chunk.start += static_cast<index_t>(size);
        ITextureBufferHandle handle;
        handle.chunkIdx = index;
        handle.start    = start;
        handle.end      = static_cast<index_t>(start + size);
        handle.texture  = chunk.texture;

        _handles.emplace_back(handle);
        return handle;
    }
    // create a new one
    auto     targetSize = static_cast<int32_t>(std::sqrt(size / _formatSize));
    uint32_t texLength  = _roundUpFn ? _roundUpFn(targetSize, _formatSize) : std::max(1024, static_cast<int>(utils::nextPOT(targetSize)));
    auto     newChunk   = _chunks[createChunk(texLength)];

    newChunk.start += static_cast<index_t>(size);
    ITextureBufferHandle texHandle;
    texHandle.chunkIdx = static_cast<index_t>(_chunkCount - 1);
    texHandle.start    = 0;
    texHandle.end      = static_cast<index_t>(size);
    texHandle.texture  = newChunk.texture;
    _handles.emplace_back(texHandle);
    return texHandle;
}

void TextureBufferPool::free(const ITextureBufferHandle &handle) {
    auto iter = std::find(_handles.begin(), _handles.end(), handle);
    if (iter != _handles.end()) {
        _chunks[handle.chunkIdx].end = handle.end;
        _handles.erase(iter);
    }
}

uint32_t TextureBufferPool::createChunk(uint32_t length) {
    uint32_t texSize = length * length * _formatSize;
    // debug(`TextureBufferPool: Allocate chunk $device->createTexture({gfx::TextureType::TEX2D,
    auto *texture = _device->createTexture({gfx::TextureType::TEX2D,
                                            gfx::TextureUsageBit::SAMPLED | gfx::TextureUsageBit::TRANSFER_DST,
                                            _format,
                                            length,
                                            length});

    ITextureBuffer chunk;
    chunk.texture        = texture;
    chunk.size           = texSize;
    chunk.start          = 0;
    chunk.end            = static_cast<index_t>(texSize);
    _chunks[_chunkCount] = chunk;
    return _chunkCount++;
}

void TextureBufferPool::update(const ITextureBufferHandle &handle, ArrayBuffer *buffer) {
    gfx::BufferDataList        buffers;
    uint8_t *                  bufferData = buffer->getData();
    gfx::BufferTextureCopyList regions;
    auto                       start = static_cast<int32_t>(handle.start / _formatSize);

    uint32_t remainSize = buffer->byteLength() / _formatSize;
    int32_t  offsetX    = start % static_cast<int32_t>(handle.texture->getWidth());
    int32_t  offsetY    = std::floor(start / handle.texture->getWidth());
    uint32_t copySize   = std::min(handle.texture->getWidth() - offsetX, remainSize);
    uint32_t begin      = 0;

    if (offsetX > 0) {
        _region0.texOffset.x      = offsetX;
        _region0.texOffset.y      = offsetY;
        _region0.texExtent.width  = copySize;
        _region0.texExtent.height = 1;
        buffers.emplace_back(bufferData + begin * _formatSize);
        regions.emplace_back(_region0);
        offsetX = 0;
        offsetY += 1;
        remainSize -= copySize;
        begin += copySize;
    }

    if (remainSize > 0) {
        _region1.texOffset.x = offsetX;
        _region1.texOffset.y = offsetY;

        if (remainSize > handle.texture->getWidth()) {
            _region1.texExtent.width  = handle.texture->getWidth();
            _region1.texExtent.height = std::floor(remainSize / handle.texture->getWidth());
            copySize                  = _region1.texExtent.width * _region1.texExtent.height;
        } else {
            copySize                  = remainSize;
            _region1.texExtent.width  = copySize;
            _region1.texExtent.height = 1;
        }

        buffers.emplace_back(bufferData + begin * _formatSize);
        regions.emplace_back(_region1);

        offsetX = 0;
        offsetY += static_cast<int32_t>(_region1.texExtent.height);
        remainSize -= copySize;
        begin += copySize;
    }

    if (remainSize > 0) {
        _region2.texOffset.x      = offsetX;
        _region2.texOffset.y      = offsetY;
        _region2.texExtent.width  = remainSize;
        _region2.texExtent.height = 1;

        buffers.emplace_back(bufferData + begin * _formatSize);
        regions.emplace_back(_region2);
    }
    _device->copyBuffersToTexture(buffers, handle.texture, regions);
}

index_t TextureBufferPool::findAvailableSpace(uint32_t size, index_t chunkIdx) const {
    auto    chunk   = _chunks[chunkIdx];
    bool    isFound = false;
    index_t start   = chunk.start;
    if ((start + size) <= chunk.size) {
        isFound = true;
    } else {
        start = 0; // try to find from head again
        std::vector<ITextureBufferHandle> handles;
        for (auto h : _handles) {
            if (h.chunkIdx == chunkIdx) {
                handles.emplace_back(h);
            }
        }
        std::sort(handles.begin(), handles.end(), [](const ITextureBufferHandle &a, const ITextureBufferHandle &b) { return a.start - b.start; });
        for (auto handle : handles) {
            if ((start + size) <= handle.start) {
                isFound = true;
                break;
            }
            start = handle.end;
        }
        if (!isFound && (start + size) <= chunk.size) {
            isFound = true;
        }
    }
    return isFound ? start : CC_INVALID_INDEX;
}

ITextureBufferHandle TextureBufferPool::mcDonaldAlloc(uint32_t size) {
    size = roundUp(size, _alignment);
    for (index_t i = 0; i < _chunkCount; ++i) {
        auto    chunk   = _chunks[i];
        bool    isFound = false;
        index_t start   = chunk.start;
        if ((start + size) <= chunk.end) {
            isFound = true;
        } else if (start > chunk.end) {
            if ((start + size) <= chunk.size) {
                isFound = true;
            } else if (size <= chunk.end) {
                // Try to find from head again.
                start       = 0;
                chunk.start = 0;
                isFound     = true;
            }
        } else if (start == chunk.end) {
            start       = 0;
            chunk.start = 0;
            chunk.end   = static_cast<index_t>(chunk.size);
            if (size <= chunk.end) {
                isFound = true;
            }
        }
        if (isFound) {
            chunk.start += static_cast<index_t>(size);

            ITextureBufferHandle handle;
            handle.chunkIdx = i;
            handle.start    = start;
            handle.end      = static_cast<index_t>(size);
            handle.texture  = chunk.texture;
            _handles.emplace_back(handle);
            return handle;
        }
    }
    // create a new one
    auto     targetSize = static_cast<int32_t>(std::sqrt(size / _formatSize));
    uint32_t texLength  = _roundUpFn ? _roundUpFn(targetSize, _formatSize) : std::max(1024, static_cast<int>(utils::nextPOT(targetSize)));
    auto     newChunk   = _chunks[createChunk(texLength)];

    newChunk.start += static_cast<index_t>(size);
    ITextureBufferHandle texHandle;
    texHandle.chunkIdx = static_cast<index_t>(_chunkCount);
    texHandle.start    = 0;
    texHandle.end      = static_cast<index_t>(size),
    texHandle.texture  = newChunk.texture;
    _handles.emplace_back(texHandle);
    return texHandle;
}

} // namespace cc
